### ISSUE TITLE
Modify softmax op to also have causal_mask as an argument

### ISSUE THREAD

**Issue by arakhmati** [tenstorrent-metal/tt-metal#4132]
Currently, `softmax` takes in `input_tensor`, `attention_mask` and `scalar` value:

where
     `input_tensor` is of shape `[batch_size, num_heads, sequence_size, sequence_size]`
     `attention_mask` is of shape `[batch_size, 1, 32, sequence_size]`

and the operation is computed as:

```
softmax(input_tensor * scalar + attention_mask)
```

and we need an extra argument `causal_mask` of shape  `[batch_size, 1, sequence_size, sequence_size]` such that the operation would compute as:

```
softmax(input_tensor * scalar + attention_mask + causal_mask)
```

**Comment by jvasilje** [comment]
@davorchap can someone help with this..?

**Comment by davorchap** [comment]
> @davorchap can someone help with this..?

Yes, @yugaoTT can help.

**Comment by yugaoTT** [comment]
@arakhmati Would it work if `attention_mask` and  `causal_mask` are pre-added before we run the kernels? so there's only one input mask, and we can save some computation time.

**Comment by arakhmati** [comment]
> @arakhmati Would it work if `attention_mask` and  `causal_mask` are pre-added before we run the kernels (pre-added on the model level) ? so there's only one input mask, and we can save some computation time. 

It will definitely work for the current case. Which is all we need for now
